File: python/relicpy/tree.py
"""Path trees — the SHAPE of a set of files, not just the list.

A leaf carries an opaque `label` rather than fixed columns, because the callers
describe different things: an indexed transcript has an event count, a PENDING file
does not exist in the index and has a state and a size.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Optional


@dataclass
class TreeNode:
    children: dict[str, "TreeNode"] = field(default_factory=dict)
    label: Optional[str] = None
    weight: int = 0
    files: int = 0


def build_tree(entries: list[dict]) -> TreeNode:
    root = TreeNode()
    for e in entries:
        parts = [p for p in e["path"].split("/") if p]
        w = int(e.get("weight") or 0)
        node = root
        root.weight += w
        root.files += 1
        for i, name in enumerate(parts):
            child = node.children.get(name)
            if child is None:
                child = TreeNode()
                node.children[name] = child
            child.weight += w
            child.files += 1
            if i == len(parts) - 1:
                child.label = e["label"]
            node = child
    return root


def render_tree(node: TreeNode, prefix: str = "", limit_per_dir: int = 8,
                emit: Callable[[str], None] = print, unit: str = "ev") -> None:
    """Directories first, then files, each group biggest first.

    A run directory is what the reader is looking for; burying it under twenty sibling
    transcripts defeats the point of drawing a tree.
    """
    kids = sorted(node.children.items(),
                  key=lambda kv: (0 if kv[1].children else 1, -kv[1].weight))
    shown, hidden = kids[:limit_per_dir], len(kids) - min(len(kids), limit_per_dir)
    for i, (name, child) in enumerate(shown):
        last = i == len(shown) - 1 and hidden == 0
        branch = "└── " if last else "├── "
        if child.children:
            w = f", {child.weight:,} {unit}" if child.weight else ""
            emit(f"{prefix}{branch}{name}/   ({child.files} file{'' if child.files == 1 else 's'}{w})")
            render_tree(child, prefix + ("    " if last else "│   "), limit_per_dir, emit, unit)
        else:
            emit(f"{prefix}{branch}{name}   {child.label}")
    if hidden > 0:
        emit(f"{prefix}└── ... and {hidden} more")


def common_prefix(paths: list[str]) -> str:
    """The longest directory prefix every path shares — what to print above the tree."""
    if not paths:
        return ""
    split = [p.split("/") for p in paths]
    first = split[0]
    i = 0
    while i < len(first) - 1 and all(len(p) > i and p[i] == first[i] for p in split):
        i += 1
    return "/".join(first[:i]) + "/"

File: src/tree.ts
/**
 * Path trees — the SHAPE of a set of files, not just the list.
 *
 * Lives apart from cli.ts because cli.ts runs a command at import time: anything that
 * imported it to reuse one function would execute the CLI as a side effect. Same
 * reason query.ts exists.
 *
 * A leaf carries an opaque `label` rather than a fixed set of columns, because the two
 * callers describe different things — an indexed transcript has an event count, a
 * PENDING file does not exist in the index at all and has a state and a size instead.
 * Baking one caller's columns into the renderer would have meant a second renderer.
 */

const fmt = (n: number) => n.toLocaleString("en-US");

export interface TreeEntry {
  path: string;     // relative to whatever root the caller prints above the tree
  label: string;    // rendered after the filename
  weight?: number;  // summed up the directories; events, bytes, whatever the caller means
}

export interface TreeNode {
  children: Map<string, TreeNode>;
  leaf?: { label: string };
  weight: number;
  files: number;
}

function newNode(): TreeNode { return { children: new Map(), weight: 0, files: 0 }; }

export function buildTree(entries: TreeEntry[]): TreeNode {
  const root = newNode();
  for (const e of entries) {
    const parts = e.path.split("/").filter(Boolean);
    const w = e.weight ?? 0;
    let node = root;
    root.weight += w; root.files++;
    for (let i = 0; i < parts.length; i++) {
      const name = parts[i];
      let child = node.children.get(name);
      if (!child) { child = newNode(); node.children.set(name, child); }
      child.weight += w;
      child.files++;
      if (i === parts.length - 1) child.leaf = { label: e.label };
      node = child;
    }
  }
  return root;
}

/**
 * Directories first, then files, each group biggest first.
 *
 * A run directory is the thing the reader is looking for; burying it under twenty
 * sibling transcripts defeats the point of drawing a tree at all.
 */
export function renderTree(
  node: TreeNode,
  prefix = "",
  limitPerDir = 8,
  emit: (line: string) => void = console.log,
  unit = "ev",
): void {
  const kids = [...node.children.entries()];
  kids.sort((a, b) => {
    const ad = a[1].children.size > 0, bd = b[1].children.size > 0;
    if (ad !== bd) return ad ? -1 : 1;
    return b[1].weight - a[1].weight;
  });
  const shown = kids.slice(0, limitPerDir);
  const hidden = kids.length - shown.length;
  shown.forEach(([name, child], i) => {
    const last = i === shown.length - 1 && hidden === 0;
    const branch = last ? "└── " : "├── ";
    if (child.children.size > 0) {
      const n = child.files;
      const w = child.weight ? `, ${fmt(child.weight)} ${unit}` : "";
      emit(`${prefix}${branch}${name}/   (${n} file${n === 1 ? "" : "s"}${w})`);
      renderTree(child, prefix + (last ? "    " : "│   "), limitPerDir, emit, unit);
    } else {
      emit(`${prefix}${branch}${name}   ${child.leaf!.label}`);
    }
  });
  if (hidden > 0) emit(`${prefix}└── ... and ${hidden} more`);
}

/** The longest directory prefix every path shares — what to print above the tree. */
export function commonPrefix(paths: string[]): string {
  if (!paths.length) return "";
  const split = paths.map(p => p.split("/"));
  const first = split[0];
  let i = 0;
  while (i < first.length - 1 && split.every(p => p[i] === first[i])) i++;
  return first.slice(0, i).join("/") + "/";
}
